server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetAllocationStatsAction.java

```diff
@@ -118,13 +125,65 @@ protected void doExecute(Task task, Request request, ActionListener<Response> li
     protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<Response> listener) throws Exception {
         // NB we are still on a transport thread here - if adding more functionality here make sure to fork to a different pool
 
-        final SubscribableListener<Map<String, NodeAllocationStats>> allocationStatsStep = request.metrics().contains(Metric.ALLOCATIONS)
-            ? SubscribableListener.newForked(allocationStatsSupplier::execute)
-            : SubscribableListener.newSucceeded(Map.of());
+        if (request.metrics().contains(Metric.ALLOCATIONS) == false) {
+            listener.onResponse(statsToResponse(Map.of(), request));
+            return;
+        }
+        // Perform a cheap check for the cached stats up front.
+        final var cachedStats = allocationStatsCache.get();
+        if (cachedStats != null) {
+            listener.onResponse(statsToResponse(cachedStats, request));
+            return;
+        }
+
+        assert task instanceof CancellableTask;
+        final var wrappedListener = ContextPreservingActionListener.wrapPreservingContext(listener, threadPool.getThreadContext());
+        final var taskListenerPair = new TaskListenerPair((CancellableTask) task, wrappedListener);
+
+        synchronized (this) {
+            if (waitingListeners != null) {
+                tasksList.add(taskListenerPair);
+                waitingListeners.addListener(wrappedListener);
+                return;
+            }
 
-        allocationStatsStep.andThenApply(
-            allocationStats -> new Response(allocationStats, request.metrics().contains(Metric.FS) ? diskThresholdSettings : null)
-        ).addListener(listener);
+            tasksList = new ArrayList<>();
+            waitingListeners = new SubscribableListener<>();
+            tasksList.add(taskListenerPair);
+            waitingListeners.addListener(ActionListener.runBefore(wrappedListener, () -> {
+                synchronized (this) {
+                    waitingListeners = null;
+                    tasksList = null;
+                }
+            }));
+        }
+
+        SubscribableListener.newForked(allocationStatsSupplier::accept)
+            .andThenApply(stats -> statsToResponse(stats, request))
+            .addListener(waitingListeners);
+    }
+
+    private Response statsToResponse(Map<String, NodeAllocationStats> stats, Request request) {
```

Comment:
Could we rephrase the execution into something that (maybe asynchronously) computes the allocation stats, and then combine that with the disk threshold settings in one place as in the original code? I.e. replace just `allocationStatsSupplier::execute` with something cancellation-sensitive?